### PR TITLE
Add minimal FSDP example

### PR DESCRIPTION
**Pull request by coreystatendet**
Add an example (originally by Garrett with small updates from me) that shows how to use torch's FSDP implementation for LLM training alongside Core API.

### CODE REVIEW

File: fsdp/minimal-fsdp/fsdp.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+) -> Generator[tuple[torch.Tensor, torch.Tensor], None, None]:
+    """
+    Fake dataloader. Yields a different set of data for each rank, and for train vs validation.
+    This data would usually come from a tokenized dataset.
+    """
+    generator = torch.Generator(device=device)
+    next_idx = 0
+    while True:
+        if next_idx == 0:
+            generator.manual_seed(42 + rank + 100000 * is_validation)
```
Comment (garrett361) [inline fsdp/minimal-fsdp/fsdp.py]: Just curious: why this if block? Why not seed immediately after creating the generator?

Also, this isn't very important, but I guess the restarting logic starts the dataloader over, rather than continuing from the last batch. I wouldn't fix this, just maybe note it in a comment.

File: fsdp/minimal-fsdp/fsdp.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    )
+
+    # Inspect the model:
+    if core_context.distributed.rank == 0:
+        print("Model before FSDP:")
+        print(model, flush=True)
+
+    # Wrap the embedding layer, the lm head, and each transformer block into its own FSDP unit:
+    auto_wrap_policy = ModuleWrapPolicy([TransformerBlock, EmbedAndEncode, LMHead])
+
```
Comment (garrett361) [inline fsdp/minimal-fsdp/fsdp.py]: I should have caught this before, but there are some small `use_amp = True` related issues we probably want to adress.

First, there is some special handling for the FSDP weight/comms/etc types when using amp, as in
```python
from torch.distributed.fsdp import MixedPrecision
fsdp_model = FSDP(model,
    mixed_precision=MixedPrecision(param_dtype=torch.bfloat16,
    reduce_dtype=torch.bfloat16), ...)
``` 
Not needed for correctness, just efficiency.

Second,  for the autocast in `get_loss` it's better practice to specify the `dtype=torch.bfloat16` arg, as in 
```python
with torch.cuda.amp.autocast(enabled=use_amp, dtype=torch.bfloat16):
        outputs = fsdp_model(inputs)
```
The default is `dtype=torch.float16` (not `bfloat16`).

Third, with `bfloat16` there should be no need for the `ShardedGradScaler`; only needed for `float16`.

Comment (garrett361) [reply fsdp/minimal-fsdp/fsdp.py]: Apologies, I should have caught these. Think I missed them when just looking at the diffs.

File: fsdp/minimal-fsdp/fsdp.py (hunk 1 above)
Comment (coreystatendet) [reply fsdp/minimal-fsdp/fsdp.py]: Goal: I'm trying to have the generator yield a recurring sequence of length `simulated_size_in_batches`.  I do this by reseeding to the same initial seed after `simulated_size_in_batches` steps, which comes from setting `next_idx = (next_idx + 1) % simulated_size_in_batches`.

I save repeating the code by having the initial seed and the reseeds use the same if condition here.

**Review by garrett361 — APPROVED** [review]
Looks good! I missed some things about mixed precision that might be worth fixing, but up to you.

File: fsdp/minimal-fsdp/fsdp.py (hunk 1 above)
Comment (garrett361) [reply fsdp/minimal-fsdp/fsdp.py]: ah, I missed the modulo, got it now

File: fsdp/minimal-fsdp/fsdp.py (hunk 2 above)
Comment (coreystatendet) [reply fsdp/minimal-fsdp/fsdp.py]: > First, there is some special handling for the FSDP weight/comms/etc types when using amp ... Not needed for correctness, just efficiency.

Good catch, wasn't aware of this; I assume `autocast` just fails to recognize / handle some of the buffers and operations used inside of FSDP without it?

> Second, for the autocast in get_loss it's better practice to specify the dtype=torch.bfloat16 arg, as in

`bfloat16` isn't supported for pre-A100 GPUs, right?  I imagine we want our example to run on any GPU, hence defaulting to `float16`.  Doing a quick Google search, looks like this can be made conditional via e.g.
```
compute_capability = torch.cuda.get_device_capability()
if compute_capability[0] < 8:
     ....
```
So we could switch to using `bfloat16` iff it's supported.

> Third, with bfloat16 there should be no need for the ShardedGradScaler; only needed for float16.

See above, but yes -- if we switch to `bfloat16` entirely or conditionally, we could omit.

Comment (garrett361) [reply fsdp/minimal-fsdp/fsdp.py]: Good call on hardware compatibility. Apparently if the hardware doesn't support `bfloat16`, [it silently falls back to `float32`?](https://discuss.pytorch.org/t/bfloat16-on-nvidia-v100-gpu/201629/4) Seem weird.

Anyway, your call on what to support there. Could also make it configurable as another hparam, `amp_dtype`?

> I assume autocast just fails to recognize / handle some of the buffers and operations used inside of FSDP without it?

Nah, it's a little different, IIUC. The actual weights used for the forwards are held in whatever precision you specify in `MixedPrecision`, or `float32` if omitted. Under autocasting, various tensors have their `dtypes` changed so that specific operations occur in either high or low precision. E.g. matmuls in low-precision and softmax in high-precision.  `MixedPrecision` affects which direction most of the casts occur in (among other things).

Like in non-FSDP weights would always be kept in high-precision and down-cast as needed. With FSDP and `MixedPrecision(param_dtype=torch.bfloat16)`, say, you also avoid those down-casts because the weights are already in the desired precision. But you then might need some extra up-casts elsewhere.

There are no explicit failures if you don't specify `MixedPrecision` w/ FSDP and run under `autocast`; just perf differences, and likely some (hopefully) small numerical differences also.

I likely got some details incorrect here, also. The FSDP + amp API isn't super well documented. Talked to one of the [OLMo](https://github.com/allenai/OLMo) researchers about exactly this topic and how it's confusing a while ago. (They are heavily FSDP based.)

Comment (garrett361) [reply fsdp/minimal-fsdp/fsdp.py]: Alternatively, if you specify `MixedPrecision` in FSDP but don't wrap in `autocast` none of the casts ever happen and you can end up running every part of the forwards in low-precision, rather than mixed.
